File: benchmarks/graph_bfs/java/Graph_bfs.java
import java.util.Random;

public class Graph_bfs {
    static final int EDGES_PER_NODE = 4;

    public static void main(String[] args) {
        int n = args.length > 0 ? Integer.parseInt(args[0]) : 500_000;

        int[] adj = new int[n * EDGES_PER_NODE];
        Random rnd = new Random(42);
        for (int i = 0; i < n; i++) {
            adj[i * EDGES_PER_NODE] = (i + 1) % n;
            for (int k = 1; k < EDGES_PER_NODE; k++) {
                adj[i * EDGES_PER_NODE + k] = rnd.nextInt(n);
            }
        }

        boolean[] visited = new boolean[n];
        int[] queue = new int[n];
        int head = 0, tail = 0;
        visited[0] = true;
        queue[tail++] = 0;
        int visitedCount = 1;

        while (head < tail) {
            int node = queue[head++];
            for (int k = 0; k < EDGES_PER_NODE; k++) {
                int nxt = adj[node * EDGES_PER_NODE + k];
                if (!visited[nxt]) {
                    visited[nxt] = true;
                    visitedCount++;
                    queue[tail++] = nxt;
                }
            }
        }

        if (visitedCount != n) {
            System.err.println("self-check failed: not all nodes reachable");
            System.exit(1);
        }

        System.out.println(visitedCount);
    }
}

File: benchmarks/graph_bfs/python/graph_bfs.py
import random
import sys

EDGES_PER_NODE = 4


def main():
    n = int(sys.argv[1]) if len(sys.argv) > 1 else 500_000

    random.seed(42)
    adj = [0] * (n * EDGES_PER_NODE)
    for i in range(n):
        adj[i * EDGES_PER_NODE] = (i + 1) % n
        for k in range(1, EDGES_PER_NODE):
            adj[i * EDGES_PER_NODE + k] = random.randrange(n)

    visited = [False] * n
    queue = [0] * n
    head = tail = 0
    visited[0] = True
    queue[tail] = 0
    tail += 1
    visited_count = 1

    while head < tail:
        node = queue[head]
        head += 1
        for k in range(EDGES_PER_NODE):
            nxt = adj[node * EDGES_PER_NODE + k]
            if not visited[nxt]:
                visited[nxt] = True
                visited_count += 1
                queue[tail] = nxt
                tail += 1

    assert visited_count == n, "self-check failed: not all nodes reachable"
    print(visited_count)


if __name__ == "__main__":
    main()
